File: Rev.0/firmware/lib/src/encoding/BER/encode.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <encoding/BER/BER.h>
#include <encoding/SSMP/SSMP.h>

slice pack_integer(const field *f);
slice pack_null(const field *f);
slice pack_OID(const field *f);
slice pack_sequence(const field *f);
slice pack_varint(const uint32_t length);

// clang-format off
const uint8_t RESPONSE[] = {
    48,  41,  2,   1,   0,  4,   6,  112, 
    117, 98,  108, 105, 99, 162, 28, 2, 
    1,   1,   2,   1,   0,  2,   1,  0,
    48,  17,  48,  15,  6,  7,   43, 6,
    167, 254, 32,  1,   1,  71,  4,  24, 
    42,  55,  120};
// clang-format on

message BER_encodex(const struct packet p) {
    struct message m = {
        .data = (uint8_t *)calloc(sizeof(RESPONSE), sizeof(uint8_t)),
        .length = sizeof(RESPONSE),
    };

    memmove(m.data, RESPONSE, m.length);

    return m;
}

slice BER_encode(const struct field f) {
    struct slice s = {
        .capacity = 0,
        .length = 0,
        .bytes = NULL,
    };

    switch (f.tag) {
    case FIELD_INTEGER:
        s = pack_integer(&f);
        break;

    case FIELD_NULL:
        s = pack_null(&f);
        break;

    case FIELD_OID:
        s = pack_OID(&f);
        break;

    case FIELD_SEQUENCE:
        s = pack_sequence(&f);
        break;
    }

    return s;
}

slice pack_integer(const field *f) {
    uint8_t buffer[8] = {0};
    int ix = 0;
    int64_t v = f->integer.value;

    do {
        buffer[ix++] = v & 0x00ff;
        v >>= 8;
    } while (v != 0 && ix < sizeof(buffer));

    slice s = {
        .capacity = ix + 2,
        .length = ix + 2,
        .bytes = (uint8_t *)calloc(ix + 2, sizeof(uint8_t)),
    };

    s.bytes[0] = 0x02;
    s.bytes[1] = (uint8_t)ix;
    for (int i = 0; i < ix; i++) {
        s.bytes[i + 2] = buffer[ix - i - 1];
    }

    return s;
}

slice pack_null(const field *f) {
    slice s = {
        .capacity = 2,
        .length = 2,
        .bytes = (uint8_t *)calloc(2, sizeof(uint8_t)),
    };

    s.bytes[0] = 0x05;
    s.bytes[1] = 0x00;

    return s;
}

slice pack_OID(const field *f) {
    slice buffer = {
        .capacity = 64,
        .length = 0,
        .bytes = (uint8_t *)calloc(64, sizeof(uint8_t)),
    };

    char *oid = strdup(f->OID.OID);
    char *saveptr;
    char *p = strtok_r(oid, ".", &saveptr);

    // ... skip root '0'
    if (p != NULL && strcmp(p, "0") == 0) {
        p = strtok_r(NULL, ".", &saveptr);
    }

    // ... special encoding for first byte (.1.3)
    if (p != NULL) {
        uint8_t b = 0;

        uint32_t v = 0;
        if (sscanf(p, "%lu", &v) == 1) {
            b = 40 * v;
        }

        if ((p = strtok_r(NULL, ".", &saveptr)) != NULL) {
            uint32_t v = 0;
            if (sscanf(p, "%lu", &v) == 1) {
                b += v;
            }
        }

        buffer.bytes[buffer.length++] = b;

        p = strtok_r(NULL, ".", &saveptr);
    }

    // ... remaining sub-OIDs
    while (p != NULL) {
        uint32_t v = 0;
        if (sscanf(p, "%lu", &v) == 1) {
            slice bytes = pack_varint(v);

            slice_append(&buffer, bytes);
            slice_free(&bytes);
        }

        p = strtok_r(NULL, ".", &saveptr);
    }

    free(oid);

    // ... copy to slice
    slice length = pack_varint(buffer.length);

    slice s = {
        .capacity = 16 + buffer.length,
        .length = 0,
        .bytes = (uint8_t *)calloc(16 + buffer.length, sizeof(uint8_t)),
    };

    s.bytes[s.length++] = 0x06;
    slice_append(&s, length);
    slice_append(&s, buffer);

    slice_free(&length);
    slice_free(&buffer);

    return s;
}

slice pack_sequence(const field *f) {
    // ... encode fields
    slice buffer = make_slice(64);

    if (f->sequence.fields != NULL) {
        field *e;

        for (int i = 0; i < f->sequence.fields->size; i++) {
            if ((e = f->sequence.fields->fields[i]) != NULL) {
                slice v = BER_encode(*e);

                slice_append(&buffer, v);
                slice_free(&v);
            }
        }
    }

    // ... copy to slice
    slice s = make_slice(16 + buffer.length);
    slice length = pack_varint(buffer.length);

    s.bytes[s.length++] = 0x30;
    slice_append(&s, length);
    slice_append(&s, buffer);

    slice_free(&length);
    slice_free(&buffer);

    return s;
}

// Encodes a uint32 varint as a big-endian varint
slice pack_varint(const uint32_t length) {
    uint8_t buffer[8] = {0};
    uint32_t v = length;
    int ix = sizeof(buffer) - 1;

    buffer[ix--] = v & 0x7f;
    v >>= 7;

    while (v != 0) {
        buffer[ix--] = 0x80 | (v & 0x7f);
        v >>= 7;
    };

    // ... copy to slice
    slice s = {
        .capacity = 16,
        .length = 0,
        .bytes = (uint8_t *)calloc(16, sizeof(uint8_t)),
    };

    ix++;
    while (ix < sizeof(buffer)) {
        s.bytes[s.length++] = buffer[ix++];
    }

    return s;
}

File: Rev.0/zero2w/rpcd/encoding/ASN.1/encode.go
package BER

import (
	"bytes"
	"slices"
)

func EncodeGetRequest(rq GetRequest) ([]byte, error) {
	packet := [][]byte{}

	if v, err := pack_integer(int64(rq.Version)); err != nil {
		return nil, err
	} else {
		packet = append(packet, v)
	}

	if v, err := pack_octets(rq.Community); err != nil {
		return nil, err
	} else {
		packet = append(packet, v)
	}

	pdu := [][]byte{}

	if v, err := pack_integer(int64(rq.RequestID)); err != nil {
		return nil, err
	} else {
		pdu = append(pdu, v)
	}

	// ... error
	if v, err := pack_integer(0); err != nil {
		return nil, err
	} else {
		pdu = append(pdu, v)
	}

	// ... error index
	if v, err := pack_integer(0); err != nil {
		return nil, err
	} else {
		pdu = append(pdu, v)
	}

	varbindlist := [][]byte{}
	varbind := [][]byte{}

	if v, err := pack_oid(rq.OID); err != nil {
		return nil, err
	} else {
		varbind = append(varbind, v)
	}

	if v, err := pack_null(); err != nil {
		return nil, err
	} else {
		varbind = append(varbind, v)
	}

	if v, err := pack_sequence(varbind...); err != nil {
		return nil, err
	} else {
		varbindlist = append(varbindlist, v)
	}

	if v, err := pack_sequence(varbindlist...); err != nil {
		return nil, err
	} else {
		pdu = append(pdu, v)
	}

	if v, err := pack_pdu(tagGetRequest, pdu); err != nil {
		return nil, err
	} else {
		packet = append(packet, v)
	}

	return pack_sequence(packet...)
}

func EncodeGetResponse(response GetResponse) ([]byte, error) {
	packet := [][]byte{}

	if v, err := pack_integer(int64(response.Version)); err != nil {
		return nil, err
	} else {
		packet = append(packet, v)
	}

	if v, err := pack_octets(response.Community); err != nil {
		return nil, err
	} else {
		packet = append(packet, v)
	}

	pdu := [][]byte{}

	if v, err := pack_integer(int64(response.RequestID)); err != nil {
		return nil, err
	} else {
		pdu = append(pdu, v)
	}

	if v, err := pack_integer(response.Error); err != nil {
		return nil, err
	} else {
		pdu = append(pdu, v)
	}

	if v, err := pack_integer(response.ErrorIndex); err != nil {
		return nil, err
	} else {
		pdu = append(pdu, v)
	}

	varbindlist := [][]byte{}
	varbind := [][]byte{}

	if v, err := pack_oid(response.OID); err != nil {
		return nil, err
	} else {
		varbind = append(varbind, v)
	}

	// ... value
	switch val := response.Value.(type) {
	case uint32:
		if v, err := pack_integer(int64(val)); err != nil {
			return nil, err
		} else {
			varbind = append(varbind, v)
		}

	case uint16:
		if v, err := pack_integer(int64(val)); err != nil {
			return nil, err
		} else {
			varbind = append(varbind, v)
		}

	case string:
		if v, err := pack_octets(string(val)); err != nil {
			return nil, err
		} else {
			varbind = append(varbind, v)
		}

	default:
		if v, err := pack_null(); err != nil {
			return nil, err
		} else {
			varbind = append(varbind, v)
		}
	}

	if v, err := pack_sequence(varbind...); err != nil {
		return nil, err
	} else {
		varbindlist = append(varbindlist, v)
	}

	if v, err := pack_sequence(varbindlist...); err != nil {
		return nil, err
	} else {
		pdu = append(pdu, v)
	}

	if v, err := pack_pdu(tagGetResponse, pdu); err != nil {
		return nil, err
	} else {
		packet = append(packet, v)
	}

	return pack_sequence(packet...)
}

func EncodeSetRequest(rq SetRequest) ([]byte, error) {
	packet := [][]byte{}

	if v, err := pack_integer(int64(rq.Version)); err != nil {
		return nil, err
	} else {
		packet = append(packet, v)
	}

	if v, err := pack_octets(rq.Community); err != nil {
		return nil, err
	} else {
		packet = append(packet, v)
	}

	pdu := [][]byte{}

	if v, err := pack_integer(int64(rq.RequestID)); err != nil {
		return nil, err
	} else {
		pdu = append(pdu, v)
	}

	// ... error
	if v, err := pack_integer(0); err != nil {
		return nil, err
	} else {
		pdu = append(pdu, v)
	}

	// ... error index
	if v, err := pack_integer(0); err != nil {
		return nil, err
	} else {
		pdu = append(pdu, v)
	}

	varbindlist := [][]byte{}
	varbind := [][]byte{}

	if v, err := pack_oid(rq.OID); err != nil {
		return nil, err
	} else {
		varbind = append(varbind, v)
	}

	// ... value
	switch val := rq.Value.(type) {
	case bool:
		if v, err := pack_bool(val); err != nil {
			return nil, err
		} else {
			varbind = append(varbind, v)
		}

	case uint8:
		if v, err := pack_integer(int64(val)); err != nil {
			return nil, err
		} else {
			varbind = append(varbind, v)
		}

	case uint16:
		if v, err := pack_integer(int64(val)); err != nil {
			return nil, err
		} else {
			varbind = append(varbind, v)
		}

	case uint32:
		if v, err := pack_integer(int64(val)); err != nil {
			return nil, err
		} else {
			varbind = append(varbind, v)
		}

	case string:
		if v, err := pack_octets(string(val)); err != nil {
			return nil, err
		} else {
			varbind = append(varbind, v)
		}

	default:
		if v, err := pack_null(); err != nil {
			return nil, err
		} else {
			varbind = append(varbind, v)
		}
	}

	if v, err := pack_sequence(varbind...); err != nil {
		return nil, err
	} else {
		varbindlist = append(varbindlist, v)
	}

	if v, err := pack_sequence(varbindlist...); err != nil {
		return nil, err
	} else {
		pdu = append(pdu, v)
	}

	if v, err := pack_pdu(tagSetRequest, pdu); err != nil {
		return nil, err
	} else {
		packet = append(packet, v)
	}

	return pack_sequence(packet...)
}

func pack_bool(bval bool) ([]byte, error) {
	var b bytes.Buffer
	var u8 uint8 = 0

	if bval {
		u8 = 1
	}

	if err := b.WriteByte(tagBoolean); err != nil {
		return nil, err
	}

	if l, err := pack_varuint(uint32(1)); err != nil {
		return nil, err
	} else if _, err := b.Write(l); err != nil {
		return nil, err
	}

	if err := b.WriteByte(u8); err != nil {
		return nil, err
	}

	return b.Bytes(), nil
}

func pack_integer(ival int64) ([]byte, error) {
	var b bytes.Buffer

	var v int64 = ival
	var u8 uint8 = 0
	for {
		u8 = uint8(v & 0x00ff)
		if err := b.WriteByte(u8); err != nil {
			return nil, err
		}

		v >>= 8
		if (v == 0) || (v == -1) {
			break
		}
	}

	// ... pad if +ve and most-significant bit is set OR -ve and most-significant bit is not set
	if ival > 0 && u8 > 0x007f {
		if err := b.WriteByte(0); err != nil {
			return nil, err
		}
	} else if ival < 0 && u8 < 0x80 {
		if err := b.WriteByte(0xff); err != nil {
			return nil, err
		}
	}

	// ... length
	if l, err := pack_varuint(uint32(b.Len())); err != nil {
		return nil, err
	} else {
		slices.Reverse(l)

		if _, err := b.Write(l); err != nil {
			return nil, err
		}
	}

	// ... tag
	if err := b.WriteByte(tagInteger); err != nil {
		return nil, err
	}

	slice := b.Bytes()

	slices.Reverse(slice)

	return slice, nil
}

func pack_octets(s string) ([]byte, error) {
	var b bytes.Buffer

	if err := b.WriteByte(tagOctetString); err != nil {
		return nil, err
	}

	if l, err := pack_varuint(uint32(len(s))); err != nil {
		return nil, err
	} else if _, err := b.Write(l); err != nil {
		return nil, err
	}

	if _, err := b.Write([]byte(s)); err != nil {
		return nil, err
	}

	return b.Bytes(), nil
}

func pack_null() ([]byte, error) {
	return []byte{tagNull, 0}, nil
}

func pack_oid(oid OID) ([]byte, error) {
	var b bytes.Buffer
	var ix = 0

	// ... skip root '0'
	if ix < len(oid) && oid[ix] == 0 {
		ix++
	}

	// ... special encoding for first byte (.1.3)
	if ix < len(oid) {
		v := uint8(40 * oid[ix])
		ix++

		if ix < len(oid) {
			v += uint8(oid[ix])
			ix++
		}

		if err := b.WriteByte(v); err != nil {
			return nil, err
		}
	}

	// ... remaining sub-OIDs
	for ix < len(oid) {
		v := oid[ix]

		if slice, err := pack_varuint(v); err != nil {
			return nil, err
		} else if _, err := b.Write(slice); err != nil {
			return nil, err
		}

		ix++
	}

	// ... copy to slice

	if length, err := pack_varuint(uint32(b.Len())); err != nil {
		return nil, err
	} else {
		slice := append([]byte{tagOID}, length...)
		slice = append(slice, b.Bytes()...)

		return slice, nil
	}
}

func pack_pdu(tag byte, pdu [][]byte) ([]byte, error) {
	var b bytes.Buffer

	var length uint32 = 0
	for _, v := range pdu {
		length += uint32(len(v))
	}

	if err := b.WriteByte(tag); err != nil {
		return nil, err
	}

	if l, err := pack_varuint(length); err != nil {
		return nil, err
	} else if _, err := b.Write(l); err != nil {
		return nil, err
	}

	for _, v := range pdu {
		if _, err := b.Write(v); err != nil {
			return nil, err
		}
	}

	return b.Bytes(), nil
}

func pack_sequence(fields ...[]byte) ([]byte, error) {
	var b bytes.Buffer

	var length uint32 = 0
	for _, v := range fields {
		length += uint32(len(v))
	}

	if err := b.WriteByte(tagSequence); err != nil {
		return nil, err
	}

	if l, err := pack_varuint(length); err != nil {
		return nil, err
	} else if _, err := b.Write(l); err != nil {
		return nil, err
	}

	for _, v := range fields {
		if _, err := b.Write(v); err != nil {
			return nil, err
		}
	}

	return b.Bytes(), nil
}

// Encodes a uint32 varint as a big-endian varint
func pack_varuint(u32 uint32) ([]byte, error) {
	var b bytes.Buffer

	v := u32

	if err := b.WriteByte(byte(v & 0x7f)); err != nil {
		return nil, err
	} else {
		v >>= 7
	}

	for v != 0 {
		if err := b.WriteByte(0x80 | uint8(v&0x7f)); err != nil {
			return nil, err
		} else {
			v >>= 7
		}
	}

	// ... reverse
	slice := b.Bytes()

	slices.Reverse(slice)
	return slice, nil
}
